File: roms/mahjong_pon_chin_kan.py
#
#	Mahjong Pon Chin Kan
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    merge = lambda even, odd : bytes([odd[i // 2] if i % 2 else even[i // 2] for i in range(len(even) * 2)])
    prg = merge(z.read('ponchina/u22.bin') + z.read('um2_1_2.u29'), z.read('um2_1_3.u42') + z.read('um2_1_4.u44'))
    gfx = merge(z.read('um2_1_8.u55') + z.read('um2_1_6.u28'), z.read('um2_1_7.u43') + z.read('um2_1_5.u20'))
    voi = z.read('um2_1_9.u56') + z.read('um2_1_10.u63')

rom = prg + gfx + voi

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/mahjong_pon_chin_kan.js
/*
 *	Mahjong Pon Chin Kan
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};
const merge = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0)), step = args.length;
	return args.forEach((a, i) => a.forEach((e, j) => array[j * step + i] = e)), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg = merge(cat(z.read('ponchina/u22.bin'), z.read('um2_1_2.u29')), cat(z.read('um2_1_3.u42'), z.read('um2_1_4.u44')));
const gfx = merge(cat(z.read('um2_1_8.u55'), z.read('um2_1_6.u28')), cat(z.read('um2_1_7.u43'), z.read('um2_1_5.u20')));
const voi = cat(z.read('um2_1_9.u56'), z.read('um2_1_10.u63'));

const rom = cat(prg, gfx, voi);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
